sdk/core/System.ClientModel/src/Options/RequestOptions.cs

```diff
@@ -0,0 +1,104 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.ClientModel.Internal;
+using System.Threading;
+
+namespace System.ClientModel.Primitives;
+
+/// <summary>
+/// Controls the end-to-end duration of the service method call, including
+/// the message being sent down the pipeline.  For the duration of pipeline.Send,
+/// this may change some behaviors in various pipeline policies and the transport.
+/// </summary>
+public class RequestOptions
+{
+    private bool _frozen;
+
+    private PipelinePolicy[]? _perCallPolicies;
+    private PipelinePolicy[]? _perTryPolicies;
+    private PipelinePolicy[]? _beforeTransportPolicies;
+
+    private readonly PipelineMessageHeaders _addHeaders;
+
+    public RequestOptions()
+    {
+        CancellationToken = CancellationToken.None;
+        ErrorOptions = ClientErrorBehaviors.Default;
+
+        _addHeaders = new PipelineRequestHeaders();
+    }
+
+    public CancellationToken CancellationToken { get; set; }
+
+    public ClientErrorBehaviors ErrorOptions { get; set; }
+
+    public void AddHeader(string name, string value)
+    {
+        Argument.AssertNotNull(name, nameof(name));
+        Argument.AssertNotNull(value, nameof(value));
+
+        AssertNotFrozen();
+
+        _addHeaders.Add(name, value);
+    }
+
+    public void AddPolicy(PipelinePolicy policy, PipelinePosition position)
+    {
+        Argument.AssertNotNull(policy, nameof(policy));
+
+        AssertNotFrozen();
+
+        switch (position)
+        {
+            case PipelinePosition.PerCall:
+                _perCallPolicies = ClientPipelineOptions.AddPolicy(policy, _perCallPolicies);
+                break;
+            case PipelinePosition.PerTry:
+                _perTryPolicies = ClientPipelineOptions.AddPolicy(policy, _perTryPolicies);
+                break;
+            case PipelinePosition.BeforeTransport:
+                _beforeTransportPolicies = ClientPipelineOptions.AddPolicy(policy, _beforeTransportPolicies);
+                break;
+            default:
+                throw new ArgumentException($"Unexpected value for position: '{position}'.");
+        }
+    }
+
+    // Set options on the message before sending it through the pipeline.
+    internal void Apply(PipelineMessage message)
+    {
+        _frozen = true;
```

Comment:
is there any value in calling `AssertNotFrozen` first?

Reply:
I think we want to be able to apply options even after it is frozen.  We just don't want it to change after it has been applied.

Reply:
Would we every want to apply twice? That is the scenario I had in mind for doing this check.

Reply:
Sure, a user could pass the same RequestOptions to either different service methods, or the same service method twice.